File: src/uxncli.jl
#=
Copyright (c) 2021
https://github.com/Ismael-VC/Varvara.jl/blob/main/CONTRIBUTORS.md

Permission to use, copy, modify, and distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE.
=#

module UxnCLI

import  Base: show

using  Dates: year, dayofyear, month, dayofmonth, hour, minute, second, dayofweek, now
using Printf: @sprintf

import Match

using  Match: @match

using ..Uxn: CPU, Stack, Memory, Device
using ..VarvaraOS: boot!, compute!

export dev_console, dev_system, dev_datetime, dev_file,
       inspect, system_talk, nil_talk, datetime_talk,
       console_talk

show(io::IO, s::Stack)::Nothing = print(io, """
Stack:\tptr: $(@sprintf("%04x", s.ptr))\tkptr: $(@sprintf("%04x", s.kptr))\terror: $(@sprintf("%04x", s.error))
$(join([@sprintf("%02x", i) for i in s.dat[0:15]], ' '))
$(join([@sprintf("%02x", i) for i in s.dat[16:31]], ' '))
$(join([@sprintf("%02x", i) for i in s.dat[32:47]], ' '))
$(join([@sprintf("%02x", i) for i in s.dat[48:63]], ' '))
""")

show(io::IO, m::Memory)::Nothing = print(io, """
Memory:\tptr: $(@sprintf("%04x", m.ptr))
$(join([@sprintf("%02x", i) for i in m.dat[0:15]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[16:31]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[32:47]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[48:63]], ' '))

$(join([@sprintf("%02x", i) for i in m.dat[64:79]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[78:95]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[94:111]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[112:127]], ' '))

$(join([@sprintf("%02x", i) for i in m.dat[128:143]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[142:159]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[158:175]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[174:191]], ' '))

$(join([@sprintf("%02x", i) for i in m.dat[190:207]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[206:223]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[224:239]], ' '))
$(join([@sprintf("%02x", i) for i in m.dat[240:255]], ' '))
""")

show(io::IO, d::Device)::Nothing = print(io, """
$(startswith(string(d.talk), '#') ? "Empty" : titlecase(split(string(d.talk), '_')[1]))\tDevice:\t$(@sprintf("%02x", d.id))
""")

show(io::IO, c::CPU)::Nothing = print(io, """
CPU: $(c.uuid)
ram.ptr: $(@sprintf("%04x", c.ram.ptr))\tdevs: $(sum(isassigned(c.devs, i) for i in 0:15))

$(join(Device[c.devs[i] for i in 0:15 if isassigned(c.devs, i)], ""))
wst:
ptr: $(@sprintf("%04x", c.wst.ptr))\tkptr: $(@sprintf("%04x", c.wst.kptr))\terror: $(@sprintf("%04x", c.wst.error))
$(join([@sprintf("%02x", i) for i in c.wst.dat[0:15]], ' '))
$(join([@sprintf("%02x", i) for i in c.wst.dat[16:31]], ' '))
$(join([@sprintf("%02x", i) for i in c.wst.dat[32:47]], ' '))
$(join([@sprintf("%02x", i) for i in c.wst.dat[48:63]], ' '))

rst:
ptr: $(@sprintf("%04x", c.rst.ptr))\tkptr: $(@sprintf("%04x", c.rst.kptr))\terror: $(@sprintf("%04x", c.rst.error))
$(join([@sprintf("%02x", i) for i in c.rst.dat[0:15]], ' '))
$(join([@sprintf("%02x", i) for i in c.rst.dat[16:31]], ' '))
$(join([@sprintf("%02x", i) for i in c.rst.dat[32:47]], ' '))
$(join([@sprintf("%02x", i) for i in c.rst.dat[48:63]], ' '))

src:
ptr: $(@sprintf("%04x", c.src.ptr))\tkptr: $(@sprintf("%04x", c.src.kptr))\terror: $(@sprintf("%04x", c.src.error))
$(join([@sprintf("%02x", i) for i in c.src.dat[0:15]], ' '))
$(join([@sprintf("%02x", i) for i in c.src.dat[16:31]], ' '))
$(join([@sprintf("%02x", i) for i in c.src.dat[32:47]], ' '))
$(join([@sprintf("%02x", i) for i in c.src.dat[48:63]], ' '))

dst:
ptr: $(@sprintf("%04x", c.dst.ptr))\tkptr: $(@sprintf("%04x", c.dst.kptr))\terror: $(@sprintf("%04x", c.dst.error))
$(join([@sprintf("%02x", i) for i in c.dst.dat[0:15]], ' '))
$(join([@sprintf("%02x", i) for i in c.dst.dat[16:31]], ' '))
$(join([@sprintf("%02x", i) for i in c.dst.dat[32:47]], ' '))
$(join([@sprintf("%02x", i) for i in c.dst.dat[48:63]], ' '))
""")

function uxn_halt(c::CPU, err::UInt8, name::AbstractString, id::Int)::Exception
  @error "Halted"
  throw(UXN_ERRORS[error](@sprintf("%s#%04x, at 0x%04x", id, c.ram.ptr)))
end

end  # module

if abspath(PROGRAM_FILE) == @__FILE__
    using .UxnCLI: main

    main()
end

File: src/uxncli.c
#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include "uxn.h"

/*
Copyright (c) 2021 Devine Lu Linvega

Permission to use, copy, modify, and distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE.
*/

#pragma mark - Core

static Device *devsystem, *devconsole;

static int
error(char *msg, const char *err)
{
	fprintf(stderr, "Error %s: %s\n", msg, err);
	return 0;
}

static void
inspect(Stack *s, char *name)
{
	Uint8 x, y;
	fprintf(stderr, "\n%s\n", name);
	for(y = 0; y < 0x04; ++y) {
		for(x = 0; x < 0x08; ++x) {
			Uint8 p = y * 0x08 + x;
			fprintf(stderr,
				p == s->ptr ? "[%02x]" : " %02x ",
				s->dat[p]);
		}
		fprintf(stderr, "\n");
	}
}

#pragma mark - Devices

static Uint8
system_dei(Device *d, Uint8 port)
{
	switch(port) {
	case 0x2: return d->u->wst.ptr;
	case 0x3: return d->u->rst.ptr;
	default: return d->dat[port];
	}
}

static void
system_deo(Device *d, Uint8 port)
{
	switch(port) {
	case 0x2: d->u->wst.ptr = d->dat[port]; break;
	case 0x3: d->u->rst.ptr = d->dat[port]; break;
	case 0xe:
		inspect(&d->u->wst, "Working-stack");
		inspect(&d->u->rst, "Return-stack");
		break;
	}
}

static void
console_deo(Device *d, Uint8 port)
{
	if(port == 0x1)
		d->vector = peek16(d->dat, 0x0);
	if(port > 0x7)
		write(port - 0x7, (char *)&d->dat[port], 1);
}

static void
file_deo(Device *d, Uint8 port)
{
	Uint8 read = port == 0xd;
	if(read || port == 0xf) {
		char *name = (char *)&d->mem[peek16(d->dat, 0x8)];
		Uint16 result = 0, length = peek16(d->dat, 0xa);
		long offset = (peek16(d->dat, 0x4) << 16) + peek16(d->dat, 0x6);
		Uint16 addr = peek16(d->dat, port - 1);
		FILE *f = fopen(name, read ? "rb" : (offset ? "ab" : "wb"));
		if(f) {
			if(fseek(f, offset, SEEK_SET) != -1)
				result = read ? fread(&d->mem[addr], 1, length, f) : fwrite(&d->mem[addr], 1, length, f);
			fclose(f);
		}
		poke16(d->dat, 0x2, result);
	}
}

static Uint8
datetime_dei(Device *d, Uint8 port)
{
	time_t seconds = time(NULL);
	struct tm *t = localtime(&seconds);
	switch(port) {
	case 0x0: return (t->tm_year + 1900) >> 8;
	case 0x1: return (t->tm_year + 1900);
	case 0x2: return t->tm_mon;
	case 0x3: return t->tm_mday;
	case 0x4: return t->tm_hour;
	case 0x5: return t->tm_min;
	case 0x6: return t->tm_sec;
	case 0x7: return t->tm_wday;
	case 0x8: return t->tm_yday >> 8;
	case 0x9: return t->tm_yday;
	case 0xa: return t->tm_isdst;
	default: return d->dat[port];
	}
}

static Uint8
nil_dei(Device *d, Uint8 port)
{
	return d->dat[port];
}

static void
nil_deo(Device *d, Uint8 port)
{
	if(port == 0x1) d->vector = peek16(d->dat, 0x0);
}

#pragma mark - Generics

static const char *errors[] = {"underflow", "overflow", "division by zero"};

int
uxn_halt(Uxn *u, Uint8 error, char *name, int id)
{
	fprintf(stderr, "Halted: %s %s#%04x, at 0x%04x\n", name, errors[error - 1], id, u->ram.ptr);
	return 0;
}

static int
console_input(Uxn *u, char c)
{
	devconsole->dat[0x2] = c;
	return uxn_eval(u, devconsole->vector);
}

static void
run(Uxn *u)
{
	Uint16 vec;
	while((!u->dev[0].dat[0xf]) && (read(0, &devconsole->dat[0x2], 1) > 0)) {
		vec = peek16(devconsole->dat, 0);
		if(!vec) vec = u->ram.ptr; /* continue after last BRK */
		uxn_eval(u, vec);
	}
}

static int
load(Uxn *u, char *filepath)
{
	FILE *f;
	if(!(f = fopen(filepath, "rb")))
		return 0;
	fread(u->ram.dat + PAGE_PROGRAM, sizeof(u->ram.dat) - PAGE_PROGRAM, 1, f);
	fprintf(stderr, "Loaded %s\n", filepath);
	return 1;
}

int
main(int argc, char **argv)
{
	Uxn u;
	int i, loaded = 0;

	if(!uxn_boot(&u))
		return error("Boot", "Failed");

	/* system   */ devsystem = uxn_port(&u, 0x0, system_dei, system_deo);
	/* console  */ devconsole = uxn_port(&u, 0x1, nil_dei, console_deo);
	/* empty    */ uxn_port(&u, 0x2, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0x3, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0x4, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0x5, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0x6, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0x7, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0x8, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0x9, nil_dei, nil_deo);
	/* file     */ uxn_port(&u, 0xa, nil_dei, file_deo);
	/* datetime */ uxn_port(&u, 0xb, datetime_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0xc, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0xd, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0xe, nil_dei, nil_deo);
	/* empty    */ uxn_port(&u, 0xf, nil_dei, nil_deo);

	for(i = 1; i < argc; ++i) {
		if(!loaded++) {
			if(!load(&u, argv[i]))
				return error("Load", "Failed");
			if(!uxn_eval(&u, PAGE_PROGRAM))
				return error("Init", "Failed");
		} else {
			char *p = argv[i];
			while(*p) console_input(&u, *p++);
			console_input(&u, '\n');
		}
	}
	if(!loaded)
		return error("Input", "Missing");

	run(&u);

	return 0;
}
